### DISCUSSION TITLE
How to use Vec<MyPyClass> as parameter a of pyfunction?

### DISCUSSION THREAD

**Discussion by tushushu** [PyO3/pyo3#2103] [Questions] [answered]
Hi there, I have a pyclass as below:
```rust
#[pyclass]
pub struct Foo {
    values: Vec<bool>,
}
```
and a function as below:
```rust
#[pyfunction]
pub fn bar(conditions: Vec<&Foo>) -> BooleanList {
    ...
}
```
The codes above cannot run, because of 
```
the trait `pyo3::FromPyObject<'_>` is not implemented for `std::vec::Vec<&Foo>`
```
I tried to implement the `FromPyObject` for Vec<&Foo>, but I cannot find a correct way. Could you give some hint how I could implement this?

**Comment by adamreichold** [comment, marked as answer]
Have you tried `Vec<Py<Foo>>`? `Py<T>` is a pointer to a pyclass `T` managed by the Python heap. If the Python code passes e.g. a `list` of `Foo` to you, then this list will actually contain pointers to `Foo` instance on the heap.

Alternatively you can pass by value, e.g. `Vec<Foo>` if that makes sense w.r.t. the semantics of `Foo`, e.g. I think this requires `Foo: Clone`, i.e. adding `#[derive(Clone)]` to `struct Foo`.

**Reply by tushushu** [reply]
Thanks so much @adamreichold , the answer helps me to solve the problem. The `Py<T>.borrow` works.
